consensus/types/src/light_client_finality_update.rs

```diff
@@ -28,43 +28,38 @@ pub struct LightClientFinalityUpdate<T: EthSpec> {
 
 impl<T: EthSpec> LightClientFinalityUpdate<T> {
     pub fn new(
-        chain_spec: ChainSpec,
-        beacon_state: BeaconState<T>,
-        block: BeaconBlock<T>,
+        chain_spec: &ChainSpec,
+        beacon_state: &BeaconState<T>,
+        block: &SignedBeaconBlock<T>,
         attested_state: &mut BeaconState<T>,
-        finalized_block: BeaconBlock<T>,
+        finalized_block: &SignedBlindedBeaconBlock<T>,
     ) -> Result<Self, Error> {
         let altair_fork_epoch = chain_spec
             .altair_fork_epoch
             .ok_or(Error::AltairForkNotActive)?;
-        if attested_state.slot().epoch(T::slots_per_epoch()) < altair_fork_epoch {
+        if beacon_state.slot().epoch(T::slots_per_epoch()) < altair_fork_epoch {
             return Err(Error::AltairForkNotActive);
         }
 
-        let sync_aggregate = block.body().sync_aggregate()?;
+        let sync_aggregate = block.message().body().sync_aggregate()?;
         if sync_aggregate.num_set_bits() < chain_spec.min_sync_committee_participants as usize {
             return Err(Error::NotEnoughSyncCommitteeParticipants);
         }
 
         // Compute and validate attested header.
         let mut attested_header = attested_state.latest_block_header().clone();
-        attested_header.state_root = attested_state.tree_hash_root();
+        attested_header.state_root = beacon_state.tree_hash_root();
```

Comment:
This needs to be the `attested_state`, and we should use the `update_tree_hash_cache` method to build the internal tree hash cache which is later used for computing the merkle proof.

```suggestion
        attested_header.state_root = attested_state.update_tree_hash_cache()?;
```

I noticed that finalization updates were failing to verify because the cache was unbuilt:

> LC error constructing finality update, error: LightClientUpdateError(BeaconStateError(TreeHashCacheNotInitialized))

This change should fix it. From a brief bit of testing on a mainnet node it seems to work. Light client finality updates are quite infrequent though, so we should probably develop some better testing infrastructure eventually (e.g. unit tests and interop tests on local testnets with light clients & other full nodes).